File: shortestworddistanceii.js
//Objective is to design a class that has one method: 'shortest'
//'Shortest' takes 2 words and finds the shortest distance between them

let words = ["practice", "makes", "perfect", "coding", "makes"]


//Design of a class using a hashmap to keep track of indices
//We use two pointers to traverse over both word arrays
//and update the min distance.

class WordDistance {
    constructor(words) {
        this.map = {}

        for (let i = 0; i < words.length; i++) {
            if (this.map[words[i]] == undefined) {
                this.map[words[i]] = [i]
            } else {
                this.map[words[i]].push(i)
            }
        }
    }

    shortest(word1, word2) {
        let array1 = this.map[word1]
        let array2 = this.map[word2]
        let min = Infinity

        let i = 0
        let j = 0

        while (i < array1.length && j < array2.length) {
            let index1 = array1[i]
            let index2 = array2[j]

            min = Math.min(min, Math.abs(index2 - index1))

            if (index1 < index2) {
                i++
            } else {
                j++
            }
        }

        return min
    }
}

let newWords = new WordDistance(words)
newWords.shortest('practice', 'makes')

File: shortestworddistanceii.java
//Java Solution

class WordDistance {

    Map<String, ArrayList<Integer>> map = new HashMap<>();
    
    public WordDistance(String[] words) {
        this.map = new HashMap<String, ArrayList<Integer>>();
        
        for (int i = 0; i < words.length; i++) {
            if (!this.map.containsKey(words[i])) {
                this.map.put(words[i], new ArrayList<Integer>());
                this.map.get(words[i]).add(i);
            } else {
                this.map.get(words[i]).add(i);
            }
        }
    }
    
    public int shortest(String word1, String word2) {
        ArrayList<Integer> arr1, arr2;
        
        arr1 = this.map.get(word1);
        arr2 = this.map.get(word2);
        int p1 = 0;
        int p2 = 0;
        int ans = Integer.MAX_VALUE;
        
        while (p1 < arr1.size() && p2 < arr2.size()) {
            int index1 = arr1.get(p1);
            int index2 = arr2.get(p2);
            
            ans = Math.min(ans, Math.abs(index2 - index1));
            if (index1 < index2) {
                p1++;
            } else {
                p2++;
            }
        }
        
        return ans;
    }
}